Avoid parallel reroutes in DiskThresholdMonitor

DaveCTurner opened the pull request:
Today the `DiskThresholdMonitor` limits the frequency with which it submits
reroute tasks, but it might still submit these tasks faster than the master can
process them if, for instance, each reroute takes over 60 seconds. This causes
a problem since the reroute task runs with priority `IMMEDIATE` and is always
scheduled when there is a node over the high watermark, so this can starve any
other pending tasks on the master.

This change avoids further updates from the monitor while its last task(s) are
still in progress, and it measures the time of each update from the completion
time of the reroute task rather than its start time, to allow a larger window
for other tasks to run.

Fixes #40174

DaveCTurner committed:
Avoid parallel reroutes in DiskThresholdMonitor

Today the `DiskThresholdMonitor` limits the frequency with which it submits
reroute tasks, but it might still submit these tasks faster than the master can
process them if, for instance, each reroute takes over 60 seconds. This causes
a problem since the reroute task runs with priority `IMMEDIATE` and is always
scheduled when there is a node over the high watermark, so this can starve any
other pending tasks on the master.

This change avoids further updates from the monitor while its last task(s) are
still in progress, and it measures the time of each update from the completion
time of the reroute task rather than its start time, to allow a larger window
for other tasks to run.

Fixes #40174

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -92,88 +99,112 @@ private void warnAboutDiskIfNeeded(DiskUsage usage) {
+        }
+
+        final ImmutableOpenMap<String, DiskUsage> usages = info.getNodeLeastAvailableDiskUsages();
+        if (usages == null) {
```
Probably best to look at this bit ignoring whitespace changes - I removed a level of indentation.

ywelsch replied:
good tip

DaveCTurner committed:
AtomicLong

DaveCTurner committed:
Off-by-one

DaveCTurner committed:
Extract var

DaveCTurner committed:
Also assert that reroutes happen again if there's still a disk over the watermark

ywelsch reviewed (commented):
I'm still not super happy that we are sending a task to be executed at priority IMMEDIATE. I would rather have this call `RoutingService`. In that case, we could also avoid this whole business of tracking whether there is already a call in progress (that's taken care of by RoutingService). WDYT?

DaveCTurner commented:
I agree on the priority thing, but the `RoutingService` still uses `HIGH` priority and doesn't offer a notification on completion to keep the frequency low. I could add such a thing if you'd like?

ywelsch commented:
> I agree on the priority thing, but the RoutingService still uses HIGH priority and doesn't offer a notification on completion to keep the frequency low. I could add such a thing if you'd like?

I think HIGH priority is ok for now. I wonder why we need the notification on completion. What does it keep the frequency low of? If we're batching calls, it's fine to have multiple pending attempts?

DaveCTurner committed:
Merge branch 'master' into 2019-06-19-avoid-parallel-rerouting-in-disk-threshold-monitor

DaveCTurner committed:
Add completion notificartion to RerouteService

DaveCTurner committed:
Inject reroute action

DaveCTurner commented:
Ok I have added to the `RoutingService` the ability to listen for completion, and adjusted the `DiskThresholdMonitor` to make use of this. @ywelsch would you take another look?

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -379,7 +379,7 @@ public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
                     logger.trace("{}, scheduling a reroute", reason);
                 }
-                routingService.reroute(reason);
+                routingService.reroute(reason, ActionListener.wrap(() -> logger.trace("{}, reroute completed", reason)));
```
this also logs the same line on an exception :/
I would prefer two different log lines, and the failure one with the exception (same for other places in this PR)

DaveCTurner replied:
![image](https://user-images.githubusercontent.com/5058284/60193659-facaf280-982f-11e9-8ecb-6a3c91652d57.png)

Fixed in ce5946b

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -92,88 +104,112 @@ private void warnAboutDiskIfNeeded(DiskUsage usage) {
-                if (nodes.contains(node) == false) {
-                    nodeHasPassedWatermark.remove(node);
-                }
+
```
assert that rerouteAction is set?

DaveCTurner replied:
Fixed in c1d6ee0

ywelsch reviewed (changes requested):
I've left two small asks. Looking good o.w.

DaveCTurner committed:
Assert not-null

DaveCTurner committed:
Rename

DaveCTurner committed:
Log exceptions on error

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/2

ywelsch reviewed (approved):
LGTM

DaveCTurner committed:
Merge branch 'master' into 2019-06-19-avoid-parallel-rerouting-in-disk-threshold-monitor
